lib/ecto/query/builder/dynamic.ex

```diff
@@ -40,14 +40,14 @@ defmodule Ecto.Query.Builder.Dynamic do
   list is not reversed. This is useful when the dynamic expression
   is given in the middle of an expression.
   """
-  def partially_expand(kind, query, %{binding: binding} = dynamic, params, count) do
-    {expr, {_binding, params, subqueries, count}} = expand(query, dynamic, {binding, params, [], count})
+  def partially_expand(kind, query, %{binding: binding} = dynamic, params, subqueries, count) do
+    {expr, {_binding, params, subqueries, count}} = expand(query, dynamic, {binding, params, subqueries, count})
 
-    if subqueries != [] do
+    if subqueries != [] and kind != :select do
```

Comment:
I think now, instead of raising here, perhaps we should raise on the callers to `partially_expand`?

Reply:
The reason I mention this is because it feels backwards to have this function know specifically about `select`... although moving this out will lead to more duplication.

Another option is to have two `partially_expand`. One that accepts subqueries (and therefore returns subqueries) and another one without. I think I prefer this one.

Thoughts?

Reply:
As I was replying, I noticed this is actually not needed for this PR yet. I reverted the changes around `partially_expand`. Sorry, my bad 😅 Would you mind copying your comment to the other PR?

Reply:
I see. But this allows subquery inside dynamic... and perhaps we should support that on both `where` and `select`. So maybe this should be another PR to be considered before #3927?

Reply:
Alright, will extract it into another PR tomorrow.

Reply:
As far as I can see, subquery inside dynamic is already supported on `where`:

```elixir
subquery =
  from(t in ListTemplate,
    where: parent_as(^:list).from_template_id == t.id,
    select: %{title: t.title}
  )

where = dynamic([l], subquery(subquery) == l.title)
query = from(l in List, as: :list, where: ^where)

assert [%{title: "FANCY TEMPLATE"}] = Repo.all(query)
```

What am I missing?

Reply:
Yes, it works for where! The idea is to provide the same for select!

Reply:
But there currently is no dynamic for select, that's what #3927 would introduce.

Do you mean creating a separate PR for introducing dynamic for select at root level only, extracted from #3927? 🤔